User endpoints

adaszkiewicza opened the pull request.

Commit:
backend: add user delete endpoint + test

Commit:
backend: add user get endpoint + tests + migration

Commit:
backend: add images to delete and get user

Commit:
backend: add /user put + test

Commit:
backend - add search endpoint looking for users

KarolDenst commented on backend/YouTubeV2.Application/Validator/UserIDValidator.cs:
```diff
@@ -0,0 +1,25 @@
+            RuleFor(id => id).NotNull().NotEmpty()
+                .MustAsync(async (id, cancellationToken) => await userManager.FindByIdAsync(id) != null)
+                .WithMessage("User with provided ID does not exist")
+                .Must(IsGuid)
```
I don't think this is needed since if we find the user id it must be a guid anyways

adaszkiewicza replied:
done

KarolDenst commented on backend/YouTubeV2/Controllers/SearchController.cs:
```diff
@@ -0,0 +1,30 @@
+        [HttpGet("/search")]
+        public async Task<ActionResult<SearchResultsDTO>> SearchAsync([FromQuery][Required] string query, 
+            [FromQuery][Required] SortingTypes sortingCriterion, [FromQuery][Required] SortingDirections sortingType,
+            [FromQuery] DateTime beginDate, [FromQuery] DateTime endDate, 
```
I don't think beginDate or endDate are in the specification

adaszkiewicza replied:
i think they are (api v1.0.2 - the most recent I hope) [https://app.swaggerhub.com/apis-docs/MATEUSZZAGORSKI2/VideIO/1.0.2?fbclid=IwAR3COWVYxAgUmL1NxOukYN1r1bqvCMrgUDfVeog2ID5K_2_w93o9eePKiFA]

KarolDenst commented on backend/YouTubeV2/Controllers/SearchController.cs:
```diff
@@ -0,0 +1,30 @@
+            [FromQuery] DateTime beginDate, [FromQuery] DateTime endDate, 
+            CancellationToken cancellationToken)
+        {
+            SearchResultsDTO result = await _searchService.SearchAsync(query, sortingType, 
```
you can just return it. no need to create result

adaszkiewicza replied:
done

KarolDenst commented on backend/YouTubeV2/Controllers/UserController.cs:
```diff
@@ -21,5 +24,29 @@ public async Task<IActionResult> RegisterAsync([FromBody] RegisterDto registerDt
+        }
+
+        [HttpGet("/user")]
+        public async Task<ActionResult<UserDTO>> GetAsync([FromQuery] string userID, CancellationToken cancellationToken)
```
I think userId should be a Guid not string but I'm not sure

adaszkiewicza replied:
done

KarolDenst commented on frontend/angular.json:
```diff
@@ -2,7 +2,8 @@
   "cli": {
-    "packageManager": "npm"
+    "packageManager": "npm",
+    "analytics": false
```
what is this?

adaszkiewicza replied:
oops, will remove

KarolDenst commented on backend/YouTubeV2.Application/Model/User.cs:
```diff
@@ -5,9 +5,11 @@ namespace YouTubeV2.Application.Model
-
+        public string Name { get; set; }
+        public string Surname { get; set; }
+        public decimal AccountBalance { get; set; }
```
there is no change in EntityConfiguration for user. Should probably add these parameters there

adaszkiewicza replied:
i am not sure why. I couldn't find any methods that make sense (HasMxxLength makes no senes here). Should I just add sth like 
```c#
builder.Property(x => x.AccountBallance);
```
? Or sth else? (help appreciated)

KarolDenst commented on backend/YouTubeV2.Application/Services/SearchService.cs:
```diff
@@ -0,0 +1,39 @@
+            SortingTypes sortingType, DateTime dateBegin, DateTime dateEnd, CancellationToken cancellationToken)
+        {
+            if (query.IsNullOrEmpty())
+                throw new BadRequestException(new ErrorResponseDTO("Query cannot be empty"));
```
Just throw BadRequestException. Middleware will create ErrorResponseDTO. Also I dont think this is necessary since we check for nullorempty in validator already.

adaszkiewicza replied:
done, and changed in other places with the same issue

KarolDenst commented on backend/YouTubeV2.Application/Services/UserService.cs:
```diff
@@ -53,5 +60,178 @@ public async Task RegisterAsync(RegisterDto registerDto, CancellationToken cance
+            }
+
+            await VerifyEmailUniqueAsync(user);
+            await VerifyNameUniqueAsync(user);
```
I think these should be in validation

adaszkiewicza replied:
i agree but i am not sure i know how to do it (i need to pass a user with provided id to lambdas in `.MustAsync()` and i dont know how.  If it can be done help would be appreciated.).

KarolDenst commented on backend/YouTubeV2.Application/Services/UserService.cs:
```diff
@@ -53,5 +60,178 @@ public async Task RegisterAsync(RegisterDto registerDto, CancellationToken cance
+
+            var user = await _userManager.FindByIdAsync(userDTO.id);
+
+            if (!userDTO.avatarImage.IsNullOrEmpty())
```
I think you should delete profile image if this is Null or empty

adaszkiewicza replied:
i think you are right, done

KarolDenst commented on backend/YouTubeV2.Application/Services/UserService.cs:
```diff
@@ -53,5 +60,178 @@ public async Task RegisterAsync(RegisterDto registerDto, CancellationToken cance
+                    throw new BadRequestException(result.Errors.Select(error => new ErrorResponseDTO(error.Description)));
+            }
+        }
+        public async Task<IEnumerable<UserDTO>> SearchAsync(string query, SortingDirections sortingDirection, 
```
I would move all these methods to search service

adaszkiewicza replied:
hmm, but i use other methods form this class while performing search.  My idea was that `userService` will take care of searching users, some `videoService` of videos etc, and `searchService` would aggregate them all. But if there is a good reason for moving, I will move

KarolDenst commented on backend/YouTubeV2.Application/Services/UserService.cs:
```diff
@@ -53,5 +60,178 @@ public async Task RegisterAsync(RegisterDto registerDto, CancellationToken cance
+            return new UserDTO(user.Id, user.Email, user.UserName, user.Name, user.Surname, user.AccountBalance,
+                topRoleName, imageUri.ToString(), user.SubscriptionsCount);
+        }
+        private async Task<string> GetTopRoleForUserAsync(User user)
```
I don't like this whole top role thing. Maybe we should just edit register so that a creator only has one role. Would make things a lot easier

adaszkiewicza replied:
I agree, done

KarolDenst commented on .gitignore:
```diff
@@ -0,0 +1,8 @@
+﻿################################################################################
```
don't think this file should exist. We already have .gitignore in backed and frontend folders

adaszkiewicza replied:
added on accident, will remove

KarolDenst commented on backend/YouTubeV2.Api.Tests/SearchTests.cs:
```diff
@@ -0,0 +1,77 @@
+            foundUsers[1].nickname.Should().Be("Alaa");
+
+            response.StatusCode.Should().Be(HttpStatusCode.OK);
+
```
remove empty line

adaszkiewicza replied:
done

KarolDenst commented on backend/YouTubeV2.Api.Tests/SearchTests.cs:
```diff
@@ -0,0 +1,77 @@
+
+        }
+    }
+
```
remove empty line

adaszkiewicza replied:
done

KarolDenst commented on backend/YouTubeV2.Api.Tests/SearchTests.cs:
```diff
@@ -0,0 +1,77 @@
+                new RegisterDto("polska@polska.com", "Maikołaj", "Maikołaj", "Tomaszewski", "asdf1243@#$GJH", Role.Creator, "")
+            };
+
+            foreach (var regosterDTO in registerDtos)
```
maybe use await _webApplicationFactory.DoWithinScope<UserManager<User>> to add users to database so the tests dont rely on register endpoint working

adaszkiewicza replied:
i think i agree. Should i do the same in `UserTests`?

KarolDenst commented on backend/YouTubeV2.Api.Tests/UserTests.cs:
```diff
@@ -63,5 +65,126 @@ public async Task RegisterShouldAddToDB()
+                });
+
+            response.StatusCode.Should().Be(HttpStatusCode.OK);
+
```
remove empty line

adaszkiewicza replied:
done

KarolDenst commented on backend/YouTubeV2.Api.Tests/SearchTests.cs:
```diff
@@ -0,0 +1,77 @@
+            var searchDTO = JsonConvert.DeserializeObject<SearchResultsDTO>(responseString);
+            var foundUsers = searchDTO.users.ToList();
+
+            foundUsers.Should().HaveCount(2);
```
In the future if other tests add users to DB this will fail. We should either user Respawn to reset the database before this test or change this to check if the count is greater than 2 and users are in order in foundUsers but they dont have to be [0] and [1]

adaszkiewicza replied:
done

Commit:
fix comments

Marvvvin5 commented on backend/YouTubeV2.Api.Tests/SearchTests.cs:
```diff
@@ -0,0 +1,93 @@
+            foreach (var regosterDTO in registerDtos)
+            {
+                var stringContent = new StringContent(JsonConvert.SerializeObject(regosterDTO), Encoding.UTF8, "application/json");
+                await httpClient.PostAsync("register", stringContent);
```
I think we should add users with userManager (just like in userService) instead of using endpooint. What if endpoint doesnt work?

adaszkiewicza replied:
sure, done + fixed in non-registration UserTests

Marvvvin5 commented on backend/YouTubeV2.Api.Tests/SearchTests.cs:
```diff
@@ -0,0 +1,93 @@
+                {
+                    foreach (var registerDto in registerDtos) 
+                    {
+                        User user = new User(registerDto);
```
var

adaszkiewicza replied:
ok

Marvvvin5 commented on backend/YouTubeV2.Api.Tests/SearchTests.cs:
```diff
@@ -0,0 +1,93 @@
+            var querys = new Dictionary<string, string?>()
+            {
+                { "query", "ala" },
+                { "sortingCriterion", "Alphabetical" },
```
probably we should have a dto here with the data. Also consider adding enums like SortingCriterion.

Marvvvin5 replied:
idk if enums work with the specification and sending strings

Marvvvin5 replied:
I see you have those enums so just use them

adaszkiewicza replied:
yaml look identical so i think its ok like this. Will use enums here

Marvvvin5 commented on backend/YouTubeV2.Api.Tests/UserTests.cs:
```diff
@@ -63,5 +65,124 @@ public async Task RegisterShouldAddToDB()
+                });
+
+            response.StatusCode.Should().Be(HttpStatusCode.OK);
+        }
```
new line

adaszkiewicza replied:
done

Marvvvin5 commented on backend/YouTubeV2.Api.Tests/UserTests.cs:
```diff
@@ -63,5 +65,124 @@ public async Task RegisterShouldAddToDB()
+                });
+
+            response.StatusCode.Should().Be(HttpStatusCode.OK);
+        }
```
new line

adaszkiewicza replied:
done

Marvvvin5 commented on backend/YouTubeV2.Application/Model/User.cs:
```diff
@@ -5,9 +5,11 @@ namespace YouTubeV2.Application.Model
+        public string Surname { get; set; }
+        public decimal AccountBalance { get; set; }
+        public int SubscriptionsCount { get; set; }
+        public DateTime AccountCreationDate { get; init; }
```
I think we should always use DateTimeOffset instead of DateTime but you can ignore it

adaszkiewicza replied:
yea, DateTimeOffset is better, changed

Marvvvin5 commented on backend/YouTubeV2.Application/Model/User.cs:
```diff
@@ -5,9 +5,11 @@ namespace YouTubeV2.Application.Model
+        public string Name { get; set; }
+        public string Surname { get; set; }
+        public decimal AccountBalance { get; set; }
+        public int SubscriptionsCount { get; set; }
```
having subscriptionCount is a bad idea. We will get user's sub count based on the list of the subscriptions

adaszkiewicza replied:
hmmm, but wouldnt sorting based on subcount take a long time like this?

Marvvvin5 commented on backend/YouTubeV2.Application/Services/SearchService.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            return await _userService.SearchAsync(query, sortingDirection, sortingType, dateBegin, dateEnd, cancellationToken);
+        }
+        
```
delete empty line

adaszkiewicza replied:
done

Marvvvin5 commented on backend/YouTubeV2/Properties/serviceDependencies.local.json:
```diff
@@ -0,0 +1,16 @@
+{
```
delete

adaszkiewicza replied:
done

Marvvvin5 commented on backend/YouTubeV2/Properties/serviceDependencies.json:
```diff
@@ -0,0 +1,12 @@
+{
```
delete

adaszkiewicza replied:
done

Marvvvin5 commented on backend/YouTubeV2/Controllers/SearchController.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace YouTubeV2.Api.Controllers
+{
+    [AllowAnonymous]
```
should be [Authorize(Roles = Role.Simple)]

adaszkiewicza replied:
will remove for now and add once login is merged

Marvvvin5 commented on backend/YouTubeV2/.config/dotnet-tools.json:
```diff
@@ -0,0 +1,12 @@
+{
```
delete

Marvvvin5 commented on backend/YouTubeV2.Application/Validator/UserDTOValidator.cs:
```diff
@@ -0,0 +1,35 @@
+        {
+            RuleFor(x => x.name).NotNull().Length(1, UserConstants.MaxUserNameLength);
+            RuleFor(x => x.surname).NotNull().Length(1, UserConstants.MaxUserSurnameLength);
+            RuleFor(x => x.userType).Must(userType => userType.Equals(Role.Simple, StringComparison.InvariantCultureIgnoreCase)
```
what about administrator? Whats the validation message? If we are using must we need custom message

adaszkiewicza replied:
i dont think there should be an option to give a user an admin account - i thought we would just add admins to DB ourselves.  Message added. Also added the same message in `RegisterDtoValidator`.

Marvvvin5 commented on backend/YouTubeV2.Application/Services/UserService.cs:
```diff
@@ -32,26 +39,180 @@ public async Task RegisterAsync(RegisterDto registerDto, CancellationToken cance
+
+            if (registerDto.avatarImage.IsNullOrEmpty()) return;
+
+            var newUser = await _userManager.FindByEmailAsync(registerDto.email);
```
do we use newUser?

adaszkiewicza replied:
fixed

Marvvvin5 commented on backend/YouTubeV2.Application/Services/AzureServices/BlobServices/BlobImageService.cs:
```diff
@@ -37,5 +40,15 @@ private async Task UploadImageAsync(byte[] bytes, string fileName, string blobCo
                 cancellationToken);
         }
+
+        private async Task DeleteImageAsync(string fileName, string blobContainerName, CancellationToken cancellationToken = default)
```
idk if thats necessary. I think that uploading an image with the same name will overwrite the old image. You can check if thats true

adaszkiewicza replied:
we still should delete when deleting a user. Or do you think about uploading an empty image when deleting?

Commit:
some more fixes

Remove tests depending on one another, small migration (money + DateTimeOffset), remove accidental files, move search logic to search service

Commit:
remove unused namespaces
